From ac57205ebfdfd536962504454f9076ffd8b59d15 Mon Sep 17 00:00:00 2001
From: Torsten Landschoff <torsten.landschoff@dynamore.de>
Date: Thu, 29 Jul 2010 17:20:00 +0200
Subject: [PATCH] Add support for undetermined state for checkboxes in wxRendererGTK.

Also adds the undetermined state to the render sample so that one
can see the difference.
---
 samples/render/render.cpp |    9 ++++++++-
 src/gtk/renderer.cpp      |   11 ++++++++++-
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/samples/render/render.cpp b/samples/render/render.cpp
index c583377..db50a56 100644
--- a/samples/render/render.cpp
+++ b/samples/render/render.cpp
@@ -124,6 +124,8 @@ private:
         { OnToggleDrawFlag(event, wxCONTROL_CHECKED); }
     void OnDrawHot(wxCommandEvent& event)
         { OnToggleDrawFlag(event, wxCONTROL_CURRENT); }
+    void OnDrawUndetermined(wxCommandEvent &event)
+        { OnToggleDrawFlag(event, wxCONTROL_UNDETERMINED); }
 
     void OnAlignLeft(wxCommandEvent& WXUNUSED(event))
         { OnChangeAlign(wxALIGN_LEFT); }
@@ -195,6 +197,8 @@ private:
             flagsString += "wxCONTROL_CURRENT ";
         if ( m_flags & wxCONTROL_CHECKED )
             flagsString += "wxCONTROL_CHECKED ";
+        if ( m_flags & wxCONTROL_UNDETERMINED )
+            flagsString += "wxCONTROL_UNDETERMINED ";
         if ( flagsString.empty() )
             flagsString = "(none)";
         dc.DrawText("Using flags: " + flagsString, x1, y);
@@ -291,6 +295,7 @@ enum
     Render_DrawPressed,
     Render_DrawChecked,
     Render_DrawHot,
+    Render_DrawUndetermined,
 
     Render_AlignLeft,
     Render_AlignCentre,
@@ -326,7 +331,7 @@ BEGIN_EVENT_TABLE(MyFrame, wxFrame)
     EVT_MENU(Render_DrawPressed, MyFrame::OnDrawPressed)
     EVT_MENU(Render_DrawChecked, MyFrame::OnDrawChecked)
     EVT_MENU(Render_DrawHot, MyFrame::OnDrawHot)
-
+    EVT_MENU(Render_DrawUndetermined, MyFrame::OnDrawUndetermined)
     EVT_MENU(Render_AlignLeft, MyFrame::OnAlignLeft)
     EVT_MENU(Render_AlignCentre, MyFrame::OnAlignCentre)
     EVT_MENU(Render_AlignRight, MyFrame::OnAlignRight)
@@ -405,6 +410,8 @@ MyFrame::MyFrame()
                               "Draw in &checked state\tCtrl-C");
     menuFile->AppendCheckItem(Render_DrawHot,
                               "Draw in &hot state\tCtrl-H");
+    menuFile->AppendCheckItem(Render_DrawUndetermined,
+                              "Draw in unde&termined state\tCtrl-T");
     menuFile->AppendSeparator();
 
     menuFile->AppendRadioItem(Render_AlignLeft, "&Left align\tCtrl-1");
diff --git a/src/gtk/renderer.cpp b/src/gtk/renderer.cpp
index d859f68..c905ae8 100644
--- a/src/gtk/renderer.cpp
+++ b/src/gtk/renderer.cpp
@@ -469,12 +469,21 @@ wxRendererGTK::DrawCheckBox(wxWindow* win,
     else
         state = GTK_STATE_NORMAL;
 
+    GtkShadowType shadow_type;
+
+    if ( flags & wxCONTROL_UNDETERMINED )
+        shadow_type = GTK_SHADOW_ETCHED_IN;
+    else if ( flags & wxCONTROL_CHECKED )
+        shadow_type = GTK_SHADOW_IN;
+    else
+        shadow_type = GTK_SHADOW_OUT;
+
     gtk_paint_check
     (
         button->style,
         gdk_window,
         state,
-        flags & wxCONTROL_CHECKED ? GTK_SHADOW_IN : GTK_SHADOW_OUT,
+        shadow_type,
         NULL,
         button,
         "cellcheck",
-- 
1.6.3.3

